### DISCUSSION TITLE
【腾讯犀牛鸟2026】pnnx for torch exported program

### DISCUSSION THREAD

**Discussion by magician336** [Tencent/ncnn#6992] [Show and tell]
## 摘要

本课题为 pnnx 增加了读取和转换 `torch.export.save()` 所生成 `.pt2` 模型的能力。用户无需为了使用 ncnn 退回 TorchScript tracing，可以直接沿 `.pt2 → pnnx IR → ncnn param/bin` 完成转换。

本课题完成了 `.pt2` 格式调研、纯 C++ 容器与 schema 解析、图和权重转写、PT2 形态归一化、默认参数恢复、回归测试及测试接入，相关实现已通过 PR [Tencent/ncnn#6953](https://github.com/Tencent/ncnn/pull/6953) 提交给社区。

目前 fork 提交 `2fc0ecf` 上的 219 个双路径场景完整结构回归为 **PASS 207 / DIFF 9 / EXPORT_FAIL 1 / SKIP 2 / PNNX_PT2_FAIL 0**；PT2 专项 CTest **8/8** 通过，pyncnn 数值对拍 **10/10** 通过，Windows MSVC + LibTorch 相关 CTest **4/4** 通过。

| 项目 | 当前状态 |
|---|---|
| 主 PR | [Tencent/ncnn#6953](https://github.com/Tencent/ncnn/pull/6953)，Open |
| 开发分支 | `magician336/ncnn:pnnx-pt2-support` |
| PR 可见检查 | 开源扫描与 `license/cla` 均为 Success |
| 开发周期 | 2026-08-24 至今 |

---

## 一、问题与目标

pnnx 原有主路径面向 TorchScript `.pt`，而 PyTorch 正在推进 `torch.export`，`torch.export.save()` 的产物正是 `.pt2`。如果转换前仍要额外执行 tracing，不仅增加了用户的操作步骤，也绕开了 PyTorch 正在发展的导出接口。

本课题并非另写一套转换器，而是让 exported program 中的图、输入输出、权重、buffer 和常量进入既有 pnnx IR，继续复用现有 pass 和 ncnn lowering。实现遵守三条约束：不根据扩展名猜模型，不引入新的第三方库，也不在信息已经丢失时用启发式方法伪造语义。

有一点容易混淆：pnnx 原有的 TorchScript 路径本就使用 libtorch，本课题并未新增这项依赖。不过，`.pt2` 前端本身——包括 ZIP/JSON/schema 解析、loader 和 PT2 pass——不包含 torch/libtorch 头文件，可以独立编译测试。课题关注的是模型转换，并不试图解决 `torch.export` 自身无法导出的模型。

---

## 二、先把文件事实弄清楚

着手实现前，我先用真实导出模型和批量语料核对 `.pt2` 的实际内容，没有照搬旧格式经验：

- `.pt2` 是 ZIP 容器，权重和常量以 `ZIP_STORED` 裸字节保存；
- 计算图位于 `<root>/models/model.json`；
- config JSON 给出 dtype、shape、stride 和 storage offset 等元数据；
- `graph.tensor_values` 记录输入及中间张量的 shape/dtype；
- `torch.export` 会省略等于 ATen schema 默认值的实参。

在这些事实的基础上，我整理了格式参考，编写了导出检查脚本以及 C++/Python canonical dump 对拍工具。239 个真实 `.pt2` 文件经两套实现归一化后能够逐字对应，为后续 loader 的设计提供了可验证的基础。

早期方案曾使用 Python 桥接，并在桥接层同时完成解析、默认值补齐和语义归一化。实际验证后发现，这种做法把文件事实和图变换混在了一起，很难稳定构造现有 pass 要求的精确 IR 形态，还增加了运行时成本和维护负担。主线随后改为纯 C++ 前端。同方向的 PR #6933 只在我方独立实现完成后用于核对覆盖项，没有复制其中的代码、注释或命名。

---

## 三、核心设计：忠实转写，后置归一化

```text
.pt2 ZIP
   │
   ├─ model.json ──> pt2_schema ──> load_pt2 ──> 原始 ATen 形态
   │                                      │
   └─ weights/constants ──> Attribute ────┘
                                          │
                                          ▼
                          pass_level2 PT2 形态归一化
                                          │
                                          ▼
                               既有 pnnx / ncnn passes
                                          │
                                          ▼
                                  .ncnn.param / .ncnn.bin
```

loader 只负责如实表达归档中的节点、operand、参数、权重和元数据。ATen 名称保持原样，不根据张量形状或命名反推原始模型的写法。PT2 特有的图形则交给 `pass_level2`，转换为现有消费者能够识别的形态，让两条前端路径在后续 pass 中汇合。

各层职责也由此明确下来。模型探测使用 ZIP 签名与 `models/model.json` 双重特征；`StoreZipReader` 处理 central directory、Zip64 和大 offset；自包含的 `json.hpp` 负责 JSON 及损坏输入；schema 层解析 graph、signature 和 tensor metadata；builder 将图与属性如实写入 pnnx IR；默认值层最后按照 ATen schema 的参数名和顺序，补齐导出器省略的实参。

默认值并非在运行时查询 dispatcher，而是由脚本离线生成，形成可审计、可再生成的静态表。当前表中覆盖了 185 个实际出现的 ATen overload，近期补入了 `argmax`、`argmin`、`zeros_like` 和 Hann/Hamming 的 periodic overload。若遇到缺失的必填参数、无法编码的默认值或无效参数变体，loader 会明确报错，而不是构造一个看似成功、实际缺少 operand 的节点。

---

### 1. 文件、图和权重

实现内容包括 PT2 探测、主入口分发，以及最小 JSON parser、schema 和 builder。当前支持 user input/output、parameter、persistent/non-persistent buffer 与 tensor constant；非连续属性按照 `sizes`、`strides` 和 `storage_offset` 物化。shape/dtype 优先采用 `.pt2` 中记录的文件事实，CLI `inputshape` 仅用于补充缺失的元数据。

针对不受信归档，JSON 和 ZIP 的读取边界也做了加固。Zip64 EOCD、long comment、伪 EOCD、完整 extra field、记录数、短读和跨平台 64 位 seek/tell 都有对应保护与回归。遇到未知 `as_*` 参数、非法 metadata、负 offset、元素数溢出，以及无法安全表达的 mutation 或 dtype 时，程序会显式拒绝或给出可观察的错误，不会静默生成错误模型。

### 2. 图形归一化与默认值

PT2 pass 覆盖 linear、卷积、激活、cat/stack、split/unbind/tensor_split、池化、LayerNorm/RMSNorm、weight norm 和多种 module-form。`DEVICE`、`MEMORY_FORMAT`、scalar type 等序列化枚举会还原为现有 pass 需要的输入形态，`cuda:1` 的设备索引也不会退化为裸 `cuda`。

对 `ones_like`、`zeros_like`、Hann/Hamming window 等静态折叠，匹配前会检查 dtype、shape、溢出、kwargs 和单元素窗口等边界。`MaxPool` 的默认 `stride=None` 可以安全归一为 kernel size；半精度与 bfloat16 权重能够进入后续 ncnn lowering，零维属性也不会因为元素数被计算为零而丢失数据。

### 3. 测试链路与可维护性

除了 C++ 层的 JSON、schema、loader/pass 和 Zip64 回归，本课题还建立了 PT2/TorchScript 结构对拍、pyncnn 数值对拍、权重链路测试及 219 场景 sweep。PT2 测试中的大型 JSON fixture 已从容易被格式化工具破坏的 raw string 中拆出，并加入格式化损坏扫描与语法检查，防止自动格式化将原本可编译的测试改坏。

PyTorch 2.14 要求 C++20，因此 pnnx 会根据 Torch 版本选择标准：2.1–2.13 使用 C++17，2.14 及以上使用 C++20。PT2 专项测试的版本门槛与原有 TorchScript 矩阵相互独立，避免低版本环境影响已有测试。

---

## 五、使用示例

先使用 `torch.export` 导出并保存模型：

```python
import torch

model = MyModel().eval()
sample = torch.randn(1, 3, 224, 224)
exported_program = torch.export.export(model, (sample,))
torch.export.save(exported_program, "model.pt2")
```

构建 pnnx 后，直接将 `.pt2` 作为模型输入传入。对输入 shape 已完整记录在归档中的模型，pnnx 会优先使用归档元数据；下面的 `inputshape` 可用于指定转换时的输入形状：

```bash
cmake --build tools/pnnx/build -j$(nproc)
./tools/pnnx/build/src/pnnx model.pt2 'inputshape=[1,3,224,224]'
```

转换器会生成 `model.ncnn.param` 和 `model.ncnn.bin`，后续加载和推理方式与现有 ncnn 模型一致。对于包含动态 shape、多个输入或特定设备语义的模型，应以 `.pt2` 内记录的图和 tensor metadata 为准，不应通过随意改写 `inputshape` 来覆盖归档事实。

---

## 六、多平台 CI

pnnx 的常规快速构建在 GitHub Actions 的 Ubuntu、macOS 和 Windows runner 上执行：安装 Python 3.12 与 CPU 版 PyTorch，从源码配置并构建 pnnx。该矩阵用于尽早发现编译、CMake 和平台兼容性问题；非 Windows runner 还会运行既有的快速算子测试。

PT2 测试则使用独立的 Ubuntu job，固定 Python 3.12 和 CPU 版 PyTorch 2.13，完成 Python ncnn 安装、pnnx 源码构建，并运行 `ctest -R 'test_ncnn_pt2_' --output-on-failure`。它与原有 TorchScript 测试矩阵分离，因此 PT2 依赖或版本门槛的变化不会改变既有路径的验证环境。

此外，ncnn 主仓库保留覆盖不同系统、架构和工具链的构建工作流。PT2 前端采用标准 C++、ZIP 和 JSON 实现，不依赖平台特有接口；但 CI 通过表示对应配置的构建与测试成功，并不等同于每一种目标设备上的模型都已完成数值验证。

---

## 七、已知限制

- `torch.export` 自身无法导出的模型不属于 pnnx 的转换范围。例如 `nn.MultiheadAttention` 的一个场景会在导出阶段失败；应先检查导出是否成功，再进行转换。
- PT2 支持以已验证的 ATen 形态和测试语料为边界。遇到缺失的必填默认值、未知参数类型、无效 tensor metadata、无法安全表达的 dtype，或不安全的原地 mutation，转换器会显式报错或拒绝转换，而不会猜测语义。
- 219 场景结构对拍中仍有 9 个已知 DIFF，涉及 interpolate/Upsample、GRU/LSTM/RNN、STFT 和部分 adaptive pool。它们表示 PT2 与 TorchScript 产物的结构不同，不能直接视作数值错误；现有数值测试尚未覆盖所有这些差异。
- `torch.export` 在保存后可能将 scale 或 adaptive pool 的 `None` 轴物化为普通整数。该信息一旦未保存在归档中，loader 不会根据输入 shape 反推原始意图，因此这类场景可能保留为有解释的结构差异。

---

## 八、验证结果与边界

验证证据分为三层：同一模型的 TorchScript/PT2 输出结构对拍、pyncnn 相对 PyTorch reference 的数值对拍，以及覆盖 parser、Zip64、默认值、dtype、mutation 和重写边界的白盒回归。最近完整基线中的 8 项 PT2 CTest 包括 JSON、schema、regress、Zip64 和 4 项 Python 测试；pyncnn 10/10 覆盖 linear、卷积、grouped conv、batchnorm、layernorm、f16、bf16、conv+bn+relu、不同空间尺寸卷积及 smoke。Windows MSVC + LibTorch 构建中，JSON、schema、regress 和 Zip64 相关 CTest 4/4 通过；此前 MinGW 的独立 schema/Zip64 harness 也覆盖 offset 为 `2147487744` 的 sparse Zip64 用例。

结构 sweep 的结果不能只看数字：

| 结果 | 数量 | 含义 |
|---|---:|---|
| PASS | 207 | PT2 与 TorchScript 产物结构一致 |
| DIFF | 9 | 结构不同，不能直接等同于数值错误 |
| EXPORT_FAIL | 1 | `nn.MultiheadAttention` 在 `torch.export` 阶段失败 |
| SKIP | 2 | 两个 embedding 场景按测试清单跳过 |
| PNNX_PT2_FAIL | 0 | 没有 PT2 转换失败 |

9 个 DIFF 分别来自 interpolate/Upsample、GRU/LSTM/RNN、STFT，以及 Torch 2.14 下的三个 adaptive pool 场景。前两类与导出后展开的子图或已经物化的 output size 有关；adaptive pool 的 `None` 轴进入归档后会变成普通整数，无法再与用户显式写出的同尺寸区分。信息一旦丢失，与其根据输入 shape 猜测，不如保留有明确解释的 DIFF。现有数值对拍均已通过，但还没有覆盖全部 DIFF，所以不能把这些 DIFF 表述成“已验证数值正确”。

项目中途曾得到 `208/8/0`。后来为修复 shape、dtype 和 mutation 等正确性问题，部分不安全匹配被主动收窄。最终结论采用可复现的 `207/9/0`，不使用已经被后续修正取代的较高数字。

---

## 九、收获与后续

这项工作最重要的经验，是把“文件事实”“语义归一化”和“验证结论”分别处理。真实 `.pt2` 文件推翻了我最初关于 pickle/deflate 的猜测；归一化从 loader 移到后置 pass 后，每一次 review 修复都更容易定位、回归和说明。空列表默认值的编码就是一个典型例子：一旦它与 TorchScript 的 `None` 形态不一致，下游模式便无法匹配。修正这一根因后，许多场景随之恢复。

社区 review 也切实提高了实现质量。Zip64 字段、Windows 类型宽度、mutation 输出、半精度属性、默认值遗漏和格式化损坏等问题，很多不会出现在本地的 happy path 中。这些问题也说明，通过率并不是唯一目标；对转换器而言，更应避免的是静默生成错误模型。

接下来会继续跟进 PR #6953 的评审和检查，并在最新提交上补齐完整 CTest、219 场景 sweep 与数值对拍。剩余的结构差异，只有在归档事实或可靠提示充分时才会继续收口。也欢迎社区提供更多真实 `.pt2` 模型和边界案例。

## 致谢

感谢 腾讯犀牛鸟 和 ncnn 社区提供这次实践机会。感谢导师、维护者和 reviewer 针对方向、CI、格式、metadata 与 Zip64 等边界提出具体意见，也感谢 PyTorch 和 ncnn 的现有源码与测试提供了可验证的基线。
